### DISCUSSION TITLE
Adding plugin functions to the API

### DISCUSSION THREAD

**Discussion by DerEwige** [ACINQ/eclair#2522] [Developers]
I believe, I read somewhere it is possible to publish custom functions to the API from my plugins.
But I did not find any guide to this.

Could you point me at the correct classes and functions I have to look at?

**Comment by t-bast** [comment]
You can have a look at that PR which adds a new API endpoint: https://github.com/ACINQ/eclair/pull/2359

**Reply by DerEwige** [reply]
Thanks.
I found the needed lines in your PR, but struggle a bit to convert some of the Scala concepts to java

First I have to implement the RouteProvider interface (on top of Plugin and Runnable I already implemented)

`public class ChannelDB_Plugin implements Plugin, RouteProvider ,Runnable  {...}`

This means I have to implement the function "route"

```
	@Override
	public Function1<RequestContext, Future<RouteResult>> route(EclairDirectives eclairDirectives)
	{
		return ChannelDB_PassiveBalancer.RegisterRoutes(kitRef ,eclairDirectives);
		
	}
```
This have to return a function1 in Java I can not simply use the Route class (due to java scala bridging I guess)

Created a new class:

```
import java.io.Serializable;
import scala.runtime.AbstractFunction1;

public abstract class ChannelDB_SerializableFunction1<T1, R> extends AbstractFunction1<T1, R> implements Serializable {
	private static final long serialVersionUID = 1L;

}
```

And now have to implement this function in my plugin

```
	public static Function1<RequestContext, Future<RouteResult>> RegisterRoutes(Kit kitRef2, EclairDirectives eclairDirectives) {
		return new ChannelDB_SerializableFunction1<RequestContext, Future<RouteResult>>() {
			private static final long serialVersionUID = 1L;
			
			@Override
			public Future<RouteResult> apply(RequestContext v1) {
				
				return null;
			}	
		};		
	}

```

This is the function where I have to implement the likes of your code:


```
  def registerRoutes(kit: OfflineCommandsKit, eclairDirectives: EclairDirectives): Route = {
    import eclairDirectives._

    val close: Route = postRequest("offlineclose") { implicit t =>
      formFields(channelIdsFormParam, "scriptPubKey".as[ByteVector](binaryDataUnmarshaller).?, "preferredFeerateSatByte".as[FeeratePerByte].?, "minFeerateSatByte".as[FeeratePerByte].?, "maxFeerateSatByte".as[FeeratePerByte].?) {
        (channelIds, scriptPubKey_opt, preferredFeerate_opt, minFeerate_opt, maxFeerate_opt) =>
          val closingFeerates_opt = preferredFeerate_opt.map(preferredPerByte => {
            val preferredFeerate = FeeratePerKw(preferredPerByte)
            val minFeerate = minFeerate_opt.map(feerate => FeeratePerKw(feerate)).getOrElse(preferredFeerate / 2)
            val maxFeerate = maxFeerate_opt.map(feerate => FeeratePerKw(feerate)).getOrElse(preferredFeerate * 2)
            ClosingFeerates(preferredFeerate, minFeerate, maxFeerate)
          })
          if (scriptPubKey_opt.forall(Script.isNativeWitnessScript)) {
            val res = kit.closer.ask(ref => CloseChannels(ref, channelIds, scriptPubKey_opt, closingFeerates_opt))(Timeout(30 seconds), kit.system.scheduler.toTyped)
            complete(res)
          } else {
            reject(MalformedFormFieldRejection("scriptPubKey", "Non-segwit scripts are not allowed"))
          }
      }
    }

    close
  }
```

But right now I struggly to combine my stump with your code...
Any hints would be welcomed ^^

**Reply by t-bast** [reply]
Damn, I'm sorry I haven't done any Java in more than a decade, so I won't be able to help much...
Hopefully someone with more Java knowledge can chime in?

**Reply by pm47** [reply]
Have you had a look at the [akka-http doc](https://doc.akka.io/docs/akka-http/current/) ? They have plenty of examples in both Scala and Java.

**Reply by DerEwige** [reply]
yeah I've looked into the docu but run into some problems.

You are using akka.http.scaladsl.server.Route and this is not intended to be used from java. Intead I should be using 
akka.http.javadsl.server.Route

That did not work, so I tried to use the scaladsl version.
But I just found the correct solution.

I can just use akka.http.javadsl.server.Route and then use the methode [asScala](https://doc.akka.io/japi/akka-http/10.2.0/akka/http/javadsl/server/Route.html#asScala())() to convert it to the Scala version.

**Reply by DerEwige** [reply]
Got it working. Thank you

First I have to implement the RouteProvider interface (on top of Plugin and Runnable I already implemented)

`public class ChannelDB_Plugin implements Plugin, RouteProvider ,Runnable  {...}`

This means I have to implement the function "route"

```
	@Override
	public Function1<RequestContext, Future<RouteResult>> route(EclairDirectives eclairDirectives)
	{
		return ChannelDB_PassiveBalancer.RegisterRoutes(kitRef ,eclairDirectives);		
	}
```

I then use the javadsl to crate a route and convert it to the scala version

```
	public static  Function1<RequestContext, Future<RouteResult>> RegisterRoutes(Kit kitRef2, EclairDirectives eclairDirectives) {
		 Route Route = concat(
		    		path("saveclose", () ->
		    		post(() ->
		    		complete("{}"))));
		 return Route.asScala();
	 }
```

Obviously my Route is not yet complete, but I can already call it through the cli
